lib/email/receiver.rb

```diff
@@ -660,102 +660,97 @@ def trim_discourse_markers(reply)
       end
 
+      # extract proper sender when using mailman mailing list
+      if mail[:x_mailman_version].present?
+        address, name = Email::Receiver.extract_email_address_and_name_from_mailman(mail)
+        return address, name if address
+      end
+
       # For now we are only using the Reply-To header if the email has
       # been forwarded via Google Groups, which is why we are checking the
       # X-Original-From header too. In future we may want to use the Reply-To
       # header in more cases.
-      if mail["X-Original-From"].present?
-        if mail[:reply_to] && mail[:reply_to].errors.blank?
-          from_address, from_display_name =
-            extract_from_fields_from_header(
-              mail,
-              :reply_to,
-              comparison_headers: ["X-Original-From"],
-            )
-          return from_address, from_display_name if from_address
-        end
-      end
-
-      if mail[:from].errors.blank?
-        from_address, from_display_name = extract_from_fields_from_header(mail, :from)
-        return from_address, from_display_name if from_address
+      if mail[:x_original_from].present? && mail[:reply_to].present?
+        original_from_address, _ =
+          Email::Receiver.extract_email_address_and_name(mail[:x_original_from])
+        reply_to_address, reply_to_name =
+          Email::Receiver.extract_email_address_and_name(mail[:reply_to])
+        return reply_to_address, reply_to_name if original_from_address == reply_to_address
       end
 
-      return extract_from_address_and_name(mail.from) if mail.from.is_a? String
-
-      if mail.from.is_a? Mail::AddressContainer
-        mail.from.each do |from|
-          from_address, from_display_name = extract_from_address_and_name(from)
-          return from_address, from_display_name if from_address
-        end
-      end
-
-      nil
+      Email::Receiver.extract_email_address_and_name(mail[:from])
     rescue StandardError
       nil
     end
 
-    def extract_from_fields_from_header(mail_object, header, comparison_headers: [])
-      mail_object[header].each do |address_field|
-        from_address = address_field.address
-        from_display_name = address_field.display_name&.to_s
-
-        comparison_failed = false
-        comparison_headers.each do |comparison_header|
-          comparison_header_address = mail_object[comparison_header].to_s[/<([^>]+)>/, 1]
-          if comparison_header_address != from_address
-            comparison_failed = true
-            break
-          end
-        end
-
-        next if comparison_failed
-        next if !from_address&.include?("@")
-        return from_address&.downcase, from_display_name&.strip
-      end
+    def self.extract_email_address_and_name_from_mailman(mail)
+      list_address, _ = Email::Receiver.extract_email_address_and_name(mail[:list_post])
+      list_address, _ =
+        Email::Receiver.extract_email_address_and_name(mail[:x_beenthere]) if list_address.blank?
```

Comment:
Both `List-Post` and `X-BeenThere` headers have the email of the mailing list. We use that to determine which email address is the one from the sender.